pkg/controller/node/taint_controller.go

```diff
@@ -0,0 +1,394 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+
+	amv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/golang/glog"
+)
+
+const (
+	nodeUpdateChannelSize = 10000
+	podUpdateChannelSize  = 50000
+	retries               = 4
+)
+
+func computeTaintDifference(left []v1.Taint, right []v1.Taint) []v1.Taint {
+	result := []v1.Taint{}
+	for i := range left {
+		found := false
+		for j := range right {
+			if left[i] == right[j] {
+				found = true
+				break
+			}
+		}
+		if !found {
+			result = append(result, left[i])
+		}
+	}
+	return result
+}
+
+// copy of 'computeTaintDifference' - long live lack of generics...
+func computeTolerationDifference(left []v1.Toleration, right []v1.Toleration) []v1.Toleration {
+	result := []v1.Toleration{}
+	for i := range left {
+		found := false
+		for j := range right {
+			if left[i] == right[j] {
+				found = true
+				break
+			}
+		}
+		if !found {
+			result = append(result, left[i])
+		}
+	}
+	return result
+}
+
+type nodeUpdateItem struct {
+	oldNode   *v1.Node
+	newNode   *v1.Node
+	oldTaints []v1.Taint
+	newTaints []v1.Taint
+}
+
+type podUpdateItem struct {
+	oldPod         *v1.Pod
+	newPod         *v1.Pod
+	oldTolerations []v1.Toleration
+	newTolerations []v1.Toleration
+}
+
+type TaintController struct {
```

Comment:
I don't have not even strong, but any opinion on this naming;) Changed to NoExecuteTaintManager